happi/cli.py

```diff
@@ -392,6 +426,389 @@ def transfer(ctx, name: str, target: str):
+        sort_key = 'avg_time'
+    for stats in sorted(
+        full_stats,
+        key=lambda x: getattr(x, sort_key),
+        reverse=True,
+    ):
+        table.add_row([getattr(stats, key) for key in benchmark_sort_keys])
+    print('Benchmark output:')
+    print(table)
+    print('Benchmark completed successfully')
+
+
+@dataclasses.dataclass
+class Stats:
+    """
+    Collect and hold results from benchmark runs.
+    """
+    name: str
+    avg_time: float
+    iterations: int
+    tot_time: float
+    max_time: float
+    import_time: float
+
+    @classmethod
+    def from_search_result(
+        cls,
+        result: happi.SearchResult,
+        duration: float,
+        iterations: int,
+        wait_connected: bool,
+    ) -> Stats:
+        """
+        Create an object using a search result and store benchmarking info.
+        """
+        logger.debug(f'Checking stats for {result["name"]}')
+        if not duration and not iterations:
+            return Stats(
+                name=result["name"],
+                avg_time=0,
+                iterations=0,
+                tot_time=0,
+                max_time=0,
+                import_time=0,
+            )
+        raw_stats: List[float] = []
+        import_time = cls.import_benchmark(result)
+        if iterations > 0:
+            for _ in range(iterations):
+                raw_stats.append(cls.run_one_benchmark(
+                    result=result,
+                    wait_connected=wait_connected
+                ))
+        else:
+            start = time.monotonic()
+            while time.monotonic() - start < duration:
+                raw_stats.append(cls.run_one_benchmark(
+                    result=result,
+                    wait_connected=wait_connected
+                ))
+        return Stats(
+            name=result["name"],
+            avg_time=sum(raw_stats) / len(raw_stats),
+            iterations=len(raw_stats),
+            tot_time=sum(raw_stats),
+            max_time=max(raw_stats),
+            import_time=import_time,
+        )
+
+    @staticmethod
+    def import_benchmark(result: happi.SearchResult) -> float:
+        """
+        Check only the module import in isolation.
+        """
+        start = time.monotonic()
+        happi.loader.import_class(result.item.device_class)
+        return time.monotonic() - start
+
+    @staticmethod
+    def run_one_benchmark(
+        result: happi.SearchResult,
+        wait_connected: bool,
+    ) -> float:
+        """
+        Create one object and time it.
+        """
+        start = time.monotonic()
+        device = result.get(use_cache=False)
+        if wait_connected:
+            try:
+                device.wait_for_connection(timeout=10.0)
+            except AttributeError:
+                logger.warning(
+                    f'{result["name"]} does not have wait_for_connection.'
+                )
+            except Exception:
```

Comment:
It's good to have the broad `Exception` here, though I think we can be more specific for timeouts:
`except TimeoutError` -> current message
`except Exception` -> general failure message

Reply:
Good call